src/mesh_tools.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::HashMap;
+
+use smallvec::SmallVec;
+
+use crate::convert::{cast_u32, cast_usize};
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::face_to_face_topology;
+
+fn crawl_faces(
+    current_face_index: u32,
+    face_to_face: &HashMap<u32, SmallVec<[u32; 8]>>,
+    available_faces: &mut Vec<u32>,
+    separate_faces: &mut Vec<u32>,
+) {
+    separate_faces.push(current_face_index);
+
+    if let Some(all_neighbors) = face_to_face.get(&current_face_index) {
+        let neighbors: Vec<u32> = all_neighbors
+            .iter()
+            .copied()
+            .filter(|n| available_faces.iter().any(|a_f| a_f == n))
+            .collect();
+        let af: Vec<_> = available_faces
+            .iter()
+            .copied()
+            .filter(|f| neighbors.iter().all(|n| n != f))
+            .collect();
+        *available_faces = af;
+        for neighbor_index in neighbors {
+            crawl_faces(
+                neighbor_index,
+                face_to_face,
+                available_faces,
+                separate_faces,
+            );
+        }
+    }
+}
+
+/// Crawls the geometry to find continuous patches of geometry.
+/// Returns a vector of new separated geometries.
+#[allow(dead_code)]
+pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
+    let face_to_face = face_to_face_topology(geometry);
+    let faces = geometry.faces();
+    let vertices: Vec<_> = geometry.vertices().to_vec();
+    let normals: Vec<_> = geometry.normals().to_vec();
+
+    let mut available_faces: Vec<u32> = face_to_face.keys().map(|key| cast_u32(*key)).collect();
+
+    let mut patches: Vec<Geometry> = Vec::new();
+
+    while !available_faces.is_empty() {
+        let mut separate_faces: Vec<u32> = Vec::new();
+
+        if let Some(first_face_index) = available_faces.pop() {
+            crawl_faces(
+                first_face_index,
+                &face_to_face,
+                &mut available_faces,
+                &mut separate_faces,
+            );
+        }
+
+        let faces: Vec<_> = separate_faces
+            .iter()
+            .map(|face_index| faces[cast_usize(*face_index)])
+            .collect();
+
+        patches.push(
+            Geometry::from_faces_with_vertices_and_normals_remove_orphans(
```

Comment:
I thought about it but at the end of the day you'll need to perform the exact same logic here anyway. So reusing the existing function seemed fair.

Reply:
Okay, I can imagine renumbering the faces as you crawl over it. It will be a hell of an algorithm, though! Not sure if it is even better optimized as you might need some supplemental data structures to mark used vertices and normals.

Reply:
It's alright. I know it's better for our time to re-use. Was just wandering.

Reply:
If this is going to happen, create a Trello task for it and let's resolve it as wontfix for now.

Reply:
I actually still have some doubts if there is a more efficient way. I can imagine saving one pass through the filtered faces and copying (do we?) the vertices and normals as arguments into the function. The rest would be exactly the same, only the code will be way more complex.

Reply:
I am totally fine with the current approach, was just musing in my sickly delirium. I would even skip the trello task - if this becomes a problem, we'll see it on our profiling and fix it.